## Import Necessary Libraries

In [1]:
import datetime as dt
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings # Uyarılar
warnings.filterwarnings("ignore")

## Import Dataset

In [2]:
online_retail=pd.read_excel('../Customer Segmentation using RFM/online_retail_II.xlsx', sheet_name="Year 2010-2011")

In [3]:
df = online_retail.copy()
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom

## EDA

In [6]:
def check_df(dataframe):
    print("------------- Shape ---------------")
    print(dataframe.shape)
    print("------------ Columns ----------------")
    print(dataframe.columns)
    print("------------ Types -----------------")
    print(dataframe.dtypes)
    print("------------- Head ----------------")
    print(dataframe.head())
    print("------------- Tail -----------------")
    print(dataframe.tail())
    print("------------ Describe ----------------")
    print(dataframe.describe().T)

check_df(df)

------------- Shape ---------------
(541910, 8)
------------ Columns ----------------
Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object')
------------ Types -----------------
Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID           float64
Country                object
dtype: object
------------- Head ----------------
  Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price

#### Checking for Null Values

In [7]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

In [8]:
df.dropna(inplace=True)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 406830 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      406830 non-null  object        
 1   StockCode    406830 non-null  object        
 2   Description  406830 non-null  object        
 3   Quantity     406830 non-null  int64         
 4   InvoiceDate  406830 non-null  datetime64[ns]
 5   Price        406830 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      406830 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.9+ MB


In [12]:
df.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object')

#### Number of Unique Items in the Dataset

In [14]:
df['Description'].nunique()

3896

In [19]:
df['Description'].value_counts()

Description
WHITE HANGING HEART T-LIGHT HOLDER    2070
REGENCY CAKESTAND 3 TIER              1905
JUMBO BAG RED RETROSPOT               1662
ASSORTED COLOUR BIRD ORNAMENT         1418
PARTY BUNTING                         1416
                                      ... 
ANTIQUE RASPBERRY FLOWER EARRINGS        1
WALL ART,ONLY ONE PERSON                 1
GOLD/AMBER DROP EARRINGS W LEAF          1
INCENSE BAZAAR PEACH                     1
PINK BAROQUE FLOCK CANDLE HOLDER         1
Name: count, Length: 3896, dtype: int64

#### Rank the 5 Most Ordered Products from Most to Least

In [20]:
df.groupby("Description").agg({"Quantity":"sum"}).sort_values("Quantity", ascending=False).head(5)

Quantity
Description                                 
WORLD WAR 2 GLIDERS ASSTD DESIGNS      53215
JUMBO BAG RED RETROSPOT                45066
ASSORTED COLOUR BIRD ORNAMENT          35314
WHITE HANGING HEART T-LIGHT HOLDER     34147
PACK OF 72 RETROSPOT CAKE CASES        33409

#### Remove the Canceled Transactions From the Dataset

In [21]:
df = df[~df["Invoice"].str.contains("C", na=False)]

#### The Total Amount of Each Invoice and We Create That Column Just Multiply 'Price' and 'Quantity'

In [22]:
df["TotalPrice"] = df["Quantity"] * df["Price"]

In [23]:
df

Invoice StockCode                          Description  Quantity  \
0       536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1       536365     71053                  WHITE METAL LANTERN         6   
2       536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3       536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4       536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...        ...       ...                                  ...       ...   
541905  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
541909  581587      POST                              POSTAGE         1   

               InvoiceDate  Price  Customer ID         Country  TotalPrice  
0      2010-12-01 08:26:00   2.55      17850.0  United Kingdom       15.30  
1      2010-12-01 08:26:00   3.39      17850.0  United Kingdom       20.34  
2      2010-12-01 08:26:00   2.75      17850.0  United Kingdom       22.00  
3      2010-12-01 08:26:00   3.39      17850.0  United Kingdom       20.34  
4      2010-12-01 08:26:00   3.39      17850.0  United Kingdom       20.34  
...                    ...    ...          ...             ...         ...  
541905 2011-12-09 12:50:00   2.10      12680.0          France       12.60  
541906 2011-12-09 12:50:00   4.15      12680.0          France       16.60  
541907 2011-12-09 12:50:00   4.15      12680.0          France       16.60  
541908 2011-12-09 12:50:00   4.95      12680.0          France       14.85  
541909 2011-12-09 12:50:00  18.00      12680.0          France       18.00  

[397925 rows x 9 columns]

## Calculations for RFM Metrics

#### Determining the analysis date for the recency

In [25]:
df["InvoiceDate"].max()
today_date = dt.datetime(2011, 12, 11)
rfm = df.groupby('Customer ID').agg({'InvoiceDate': lambda date: (today_date - date.max()).days,
                                     'Invoice': lambda num: num.nunique(),
                                     'TotalPrice': lambda TotalPrice: TotalPrice.sum()})
rfm.columns=["Recency","Frequency","Monetary"]
rfm = rfm[rfm["Monetary"] > 0]
rfm.describe().T

count         mean          std   min      25%      50%      75%  \
Recency    4338.0    93.059474   100.012264  1.00   18.000   51.000   142.75   
Frequency  4338.0     4.272706     7.706221  1.00    1.000    2.000     5.00   
Monetary   4338.0  2054.270609  8989.229895  3.75  307.415  674.485  1661.74   

                 max  
Recency       374.00  
Frequency     210.00  
Monetary   280206.02

## Converting RFM Scores to Single Variable

In [26]:
#Date from customer's last purchase.The nearest date gets 5 and the furthest date gets 1.
rfm["recency_score"] = pd.qcut(rfm['Recency'], 5, labels=[5, 4, 3, 2, 1])
# Total number of purchases.The least frequency gets 1 and the maximum frequency gets 5.
rfm["frequency_score"] = pd.qcut(rfm["Frequency"].rank(method="first"), 5, labels=[1, 2, 3, 4, 5])
#Total spend by the customer.The least money gets 1, the most money gets 5.
rfm["monetary_score"]= pd.qcut(rfm["Monetary"],5,labels=[1,2,3,4,5])
rfm.head()

Recency  Frequency  Monetary recency_score frequency_score  \
Customer ID                                                               
12346.0          326          1  77183.60             1               1   
12347.0            3          7   4310.00             5               5   
12348.0           76          4   1797.24             2               4   
12349.0           19          1   1757.55             4               1   
12350.0          311          1    334.40             1               1   

            monetary_score  
Customer ID                 
12346.0                  5  
12347.0                  5  
12348.0                  4  
12349.0                  4  
12350.0                  2

In [27]:
#RFM - The value of 2 different variables that were formed was recorded as a RFM_SCORE
rfm["RFM_SCORE"] = (rfm["recency_score"].astype(str) + rfm["frequency_score"].astype(str))

## Segmenting Customers Using RFM Score

In [28]:
seg_map = {
    r'[1-2][1-2]': 'hibernating',
    r'[1-2][3-4]': 'at_Risk',
    r'[1-2]5': 'cant_loose',
    r'3[1-2]': 'about_to_sleep',
    r'33': 'need_attention',
    r'[3-4][4-5]': 'loyal_customers',
    r'41': 'promising',
    r'51': 'new_customers',
    r'[4-5][2-3]': 'potential_loyalists',
    r'5[4-5]': 'champions'
}
rfm['segment'] = rfm['RFM_SCORE'].replace(seg_map, regex=True)
rfm.head()

Recency  Frequency  Monetary recency_score frequency_score  \
Customer ID                                                               
12346.0          326          1  77183.60             1               1   
12347.0            3          7   4310.00             5               5   
12348.0           76          4   1797.24             2               4   
12349.0           19          1   1757.55             4               1   
12350.0          311          1    334.40             1               1   

            monetary_score RFM_SCORE      segment  
Customer ID                                        
12346.0                  5        11  hibernating  
12347.0                  5        55    champions  
12348.0                  4        24      at_Risk  
12349.0                  4        41    promising  
12350.0                  2        11  hibernating

## Interpretation of Descriptive Statistics of Segments

In [29]:
rfm[["segment", "Recency","Frequency","Monetary"]].groupby("segment").agg(["mean","count","max"]).round()

Recency            Frequency            Monetary        \
                       mean count  max      mean count  max     mean count   
segment                                                                      
about_to_sleep         53.0   352   72       1.0   352    2    472.0   352   
at_Risk               154.0   593  374       3.0   593    6   1085.0   593   
cant_loose            133.0    63  373       8.0    63   34   2796.0    63   
champions               6.0   633   13      12.0   633  210   6858.0   633   
hibernating           218.0  1071  374       1.0  1071    2    489.0  1071   
loyal_customers        34.0   819   72       6.0   819   63   2864.0   819   
need_attention         52.0   187   72       2.0   187    3    898.0   187   
new_customers           7.0    42   13       1.0    42    1    388.0    42   
potential_loyalists    17.0   484   33       2.0   484    3   1041.0   484   
promising              24.0    94   33       1.0    94    1    294.0    94   

                               
                          max  
segment                        
about_to_sleep         6208.0  
at_Risk               44534.0  
cant_loose            10254.0  
champions            280206.0  
hibernating           77184.0  
loyal_customers      124915.0  
need_attention        12602.0  
new_customers          3861.0  
potential_loyalists  168472.0  
promising              1758.0

## Observations

Can't Loose

In this category, there are 63 individuals. The last shopping activity occurred, on average, 133 days ago. Shopping transpires at a rate of 8, resulting in a cumulative purchase count of 63. A sum of £102.54 has been expended.

Strategic Move: 
Despite the extended interval since the most recent purchase (133 days), the overall volume of transactions remains substantial. This cohort exhibits a pattern of infrequent visits, yet engages in significant purchasing when they do. Conducting surveys among these patrons can furnish valuable insights, and initiating transformation can be achieved through the dissemination of tailored email campaigns.


<br>
<br>


Need Attention

Within this group, there exist 187 members. The last shopping endeavor took place, on average, 52 days ago. The shopping frequency stands at 2, culminating in a total of 3 purchases. An expenditure totaling £12,602 has been recorded.

Strategic Move: 
These patrons necessitate brand recall; hence, offering short-term discounts can serve as a prompt to rekindle their connection with our brand and encourage them to engage in further shopping activities.